packages/financial-templates-lib/src/price-feed/BalancerPriceFeed.js

```diff
@@ -20,47 +36,140 @@ class BalancerPriceFeed extends PriceFeedInterface {
+    }
+    return historicalPrice && this.convertDecimals(historicalPrice);
   }
   getLastUpdateTime() {
     return this.lastUpdateTime;
   }
   getCurrentPrice() {
-    // current price can be undefined, will throw for any other errors
-    return this.priceHistory.currentPrice();
+    let currentPrice;
+    // If twap window is 0, then return last price
+    if (this.twapLength === 0) {
+      currentPrice = this.getSpotPrice();
+    } else {
+      currentPrice = this._computeTwap(this.lastUpdateTime - this.twapLength, this.lastUpdateTime);
+    }
+    return currentPrice && this.convertDecimals(currentPrice);
+  }
+  // Not part of the price feed interface. Can be used to pull the balancer price at the most recent block.
+  // If `time` is undefined, return latest block price.
+  getSpotPrice(time) {
+    if (!time) {
+      return this.priceHistory.currentPrice() && this.convertDecimals(this.priceHistory.currentPrice());
+    } else {
+      // We want the block and price equal to or before this time
+      const block = this.blockHistory.getClosestBefore(time);
+      if (block == null) return null;
+      if (!this.priceHistory.has(block.timestamp)) {
+        return null;
+      }
+      return this.priceHistory.get(block.timestamp) && this.convertDecimals(this.priceHistory.get(block.timestamp));
+    }
   }
   async update() {
-    this.lastUpdateTime = await this.getTime();
+    const currentTime = await this.getTime();
+    this.logger.debug({
+      at: "BalancerPriceFeed",
+      message: "Updating",
+      lastUpdateTimestamp: currentTime
+    });
     let blocks = [];
     // disabled lookback by setting it to 0
-    if (this.lookback == 0) {
-      // handle no lookback, we just want latest block
+    if (this.lookback === 0) {
+      // handle no lookback, we just want to insert the latest block into the blockHistory.
       const block = await this.getLatestBlock();
       this.blockHistory.insert(block);
       blocks = this.blockHistory.listBlocks();
     } else {
       // handle historical lookback. Have to be careful your lookback time gives a big enough
-      // window to find a single block, otherwise you will have errors.
-      blocks = await this.blockHistory.update(this.lookback, this.lastUpdateTime);
+      // window to find a single block, otherwise you will have errors. This essentially maps
+      // blockHistory.insert() over all blocks in the lookback window.
+      blocks = await this.blockHistory.update(this.lookback + this.twapLength, currentTime);
     }
+    // The priceHistory.update() method should strip out any blocks where the price is null
     await Promise.all(blocks.map(this.priceHistory.update));
+
+    this.lastUpdateTime = currentTime;
+  }
+  // If priceHistory only encompasses 1 block, which happens if the `lookback` window is 0,
+  // then this should return the last and only price.
+  _computeTwap(startTime, endTime) {
```

Comment:
I wonder if we should make this a utility function that takes in the start time, end time, and the list of price points just so we don't have to have it in two places. The only reason I suggest it here is that the logic is somewhat tricky, so I'd prefer to only have it written in one place to avoid discrepancies. Thoughts?

Reply:
Yeah I like this I'll implement.

Reply:
Done PTAL!